### PR TITLE
Update resizing and cancellation

### PR DESCRIPTION
**Pull request by albertkol**
## Done

TLDR; List of changes:
- display messages on success or failure for both resizing and cancellation
- show up the 3DS modal if the resizing fails because of 3ds_authorizaion missing.
_______

#### [Current behaviour] Resizing works successfully 
- we refresh the page
- the new number of machines will be displayed
#### [New behaviour] Resizing works successfully
- we echo message saying that the resizing was successful
image
- we refresh the page
- the new number of machines will be displayed
_______

#### [Current behaviour] Resizing fails because of `payment_method_failed` 
- we refresh the page
- the number of machines does not change
- we show the payment methods notification
#### [New behaviour] Resizing fails because of `payment_method_failed`
- we do not refresh the page
- we show a message giving some feedback of what happened
image
- if the user tries to make any other resizing of this subscription of another subscription of the same type we will show the following message and reload the page (the reload is needed because they likely have a pending payment, and after the reload we disable the change subscription button):
image
- we show the payment methods notification after reload
_______

#### [Current behaviour] Resizing fails because of `3ds_authorisation` 
- we refresh the page
- the number of machines does not change
- we show the payment methods notification
#### [New behaviour] Resizing fails because of `3ds_authorisation` 
- we trigger the 3ds modal
##### [New behaviour] `3ds_authorisation` successful
- we echo message saying that the resizing was successful
image
- we refresh the page
- the new number of machines will be displayed
##### [New behaviour] `3ds_authorisation` failed
- we echo the following message
image
- if the user tries to make any other resizing of this subscription of another subscription of the same type we will show the following message and reload the page (the reload is needed because they likely have a pending payment, and after the reload we disable the change subscription button):
image
- we show the payment methods notification after reload
_______

#### [Current behaviour] Cancelling successful 
- we refresh the page
#### [New behaviour] Cancelling successful 
- we echo message saying that the cancellation was successful
image
- we refresh the page
#### [Current behaviour] Cancelling failed 
- page reloads, but the subscription does not get cancelled
#### [New behaviour] Cancelling failed 
- modal closes and an error message shows up: 
 - if there is a pending payment the following message shows up and we reload the page (the reload is needed because they likely have a pending payment, and after the reload we disable the change subscription button):
image
 - if there is any other error we show a more generic error message:
image

## QA

Few things to check:
1) Successful resizing up and down with normal card
- Set card number to 4242424242424242 on https://ubuntu-com-9808.demos.haus/advantage/payment-methods?test_backend=true
- Go to https://ubuntu-com-9808.demos.haus/advantage?test_backend=true
- Resize up and/or down
- You should see the Success message

2) Successful resizing up triggering 3ds modal with the card
- Set card number to 4000002760003184 on https://ubuntu-com-9808.demos.haus/advantage//payment-methods?test_backend=true
- Go to https://ubuntu-com-9808.demos.haus/advantage?test_backend=true
- Resize up
- You should trigger the 3ds modal
- Finish the 3ds check successfully
- You should see the Success message

3) Failed resizing up triggering 3ds modal with the card (4000002760003184) should show error message
- Set card number to 4000002760003184 on https://ubuntu-com-9808.demos.haus/advantage/payment-methods?test_backend=true
- Go to https://ubuntu-com-9808.demos.haus/advantage?test_backend=true
- Resize up
- You should trigger the 3ds modal
- Finish the 3ds check successfully
- You should see an error message
- try to resize again should reload page

4) Failed cancelling subscription by having a pending payment
- Have a monthly purchase
- Manually remove the `disabled` attribute from the Change Subscription button
- Click Cancel Subscription
- It should throw an error message

5) Successfully cancelling subscription
- Resolve the pending payment by going to https://ubuntu-com-9808.demos.haus/advantage/payment-methods?test_backend=true
- I will not explain how to resolve it, with the hope that the process is intuitive enough
- Once the payment is resolved go to https://ubuntu-com-9808.demos.haus/advantage?test_backend=true
- Change monthly subscription
- Cancel it
- See success message

6) Go wild try to resize both monthly and yearly subscriptions. Have multiple pending payments try to resolve them on /payment-methods

## Issue / Card

Fixes #https://github.com/canonical-web-and-design/commercial-squad/issues/3

### CODE REVIEW

**Comment by albertkol** [comment]
It started as just a job to add 3ds checks on resizing, but then we had some issues about not displaying messages and it turned into this big thing.

I _might_ have over-complicated it. So feedback is welcome, as to how to change it. One thing I noticed while I was documenting this is that cancellations should probably not be stuck because of pending payments. But we can have a chat on that too. @wgx

**Comment by ClementChaumel** [comment]
Resizing is not working correctly / not using the correct number of machines.
Here I've got 23 UAE yearly and 23 UAE monthly
image
If I resize the monthly 1 up 
image
When the page reloads I end up with 46 of them
image


------------------------------------------------------------------


Moreover, if I try to remove all my monthly but one this happens
It says it'll remove UA for 45 machines 
image
Then on reloads shows me 23 again which I suspect is my number of YEARLY UAE
image
And then if I try to downsize those 23 monthly to 1 I get this 
image

**Comment by ClementChaumel** [comment]
It's possible to get those 2 alerts to display at the same time, shouldn't they be exclusive? 
image

File: static/js/src/advantage/manage-subscription.js [hunk 1]
```diff
@@ -1,93 +1,259 @@
...
+      "<strong>Payment method:</strong> There was an error with the payment please <a href='/advantage/payment-methods'>update your payment methods</a> to fix it.",
+    subscription_missing:
+      "<strong>Cancelling subscription failed:</strong> It could be that you have a pending payment that is blocking this action. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
+    cancelling_subscription_failed:
+      "<strong>Cancelling subscription failed:</strong> Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
+    pending_purchase:
+      "<strong>Error:</strong> You already have a pending purchase. Please go to <a href='/advantage/payment-methods'>payment methods</a> to fix it. Reloading page...",
+    unknown_error:
+      "<strong>Unknown error:</strong> Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
+  };
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:24]: @wgx @elioqoshi please double-check the copy is fitting our tone

File: static/js/src/advantage/manage-subscription.js [hunk 2]
```diff
@@ -1,93 +1,259 @@
...
+      return;
+    }
+
+    handleSuccess(elements, "resizing_machines_success");
+  });
+};
+
+// It take a time for the invoice to be attached to the payment.
+// We try 5 times each 2 seconds.
+const retryGetPurchase = (purchase_id, attempts_counter, elements) => {
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:75]:
```suggestion
const retryGetPurchase = (purchaseID, attemptsCounter, elements) => {
```
Can you use camelCase, please? :camel:

File: static/js/src/advantage/manage-subscription.js [hunk 3]
```diff
@@ -1,93 +1,259 @@
...
+const requiresAuthentication = (invoice) => {
+  if (invoice.pi_decline_code === "authentication_required") {
+    return true;
+  }
+
+  if (invoice.pi_status === "requires_action" && invoice.pi_secret) {
+    return true;
+  }
+
+  return false;
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:124]:
```suggestion
return invoice.pi_decline_code === "authentication_required" || (invoice.pi_status === "requires_action" && invoice.pi_secret)
```

Since this is all we do I'm not sure we need a function here. Do you think writing the condition like that is not legible enough?

File: static/js/src/advantage/manage-subscription.js [hunk 4]
```diff
@@ -1,93 +1,259 @@
...
+      if (data.errors.includes("can only make one purchase")) {
+        handleError(elements, "pending_purchase");
+
+        setTimeout(() => {
+          location.reload();
+        }, 2000);
+
+        return;
+      }
+
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:188]:
```suggestion

       } else {

```

use else instead of return statements

File: static/js/src/advantage/manage-subscription.js [hunk 5]
```diff
@@ -1,93 +1,259 @@
...
+      if (data.errors) {
+        if (data.errors.includes("no monthly subscription")) {
+          handleError(elements, "subscription_missing");
+
+          return;
+        }
+        handleError(elements, "cancelling_subscription_failed");
+
+        return;
+      }
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:249]: Else instead of return pls :)

**Review by ClementChaumel — COMMENTED** [review]
Code is ok with a few styling nitpicks here and there :+1:

**Comment by albertkol** [comment]
@ClementChaumel resizing "problem" is a consequence of changing the number of machines in the column.

If we have bought the following:
- 2 machines of product x (monthly)
- 3 machines of product x (yearly)

The table will show now 5 machines for each row.

So when you resize, let's say the monthly one, your default value of machines in the field will be 5. You change that to 6. You are actually resizing monthly by 2 (yearly machines) + 1 (that you just added).

I faced this while testing. Obviously it's a big problem, but it does not happen on prod at the moment since we do not have monthly products yet.

Should be fixed as a separate issue.

File: static/js/src/advantage/manage-subscription.js (hunk 3 above)
Comment (albertkol) [reply static/js/src/advantage/manage-subscription.js:124]: I like this one. I do think we should keep the function.

File: static/js/src/advantage/manage-subscription.js (hunk 4 above)
Comment (albertkol) [reply static/js/src/advantage/manage-subscription.js:188]: One day we will sit down and how a chat how `else` is the root of everything evil in the IT world, causing pointless indentation and nesting. But in this case it does make the code look better. I guess JS looks so bad even an `else` can makes it prettier. :D

I am kidding, but I wish I could just to `return handleError()` so my quick returns looks nicer. I left my quick return only inside the `getPurchase` request where I think makes the code better, but if you think I should change it there, let me know.

File: static/js/src/advantage/manage-subscription.js (hunk 2 above)
Comment (albertkol) [reply static/js/src/advantage/manage-subscription.js:75]: I did not know before today that JS is into :camel:

File: static/js/src/advantage/manage-subscription.js [hunk 6]
```diff
@@ -1,93 +1,252 @@
...
+  resizeContract,
+  cancelContract,
+  getPurchase,
+} from "./contracts-api.js";
+
+const stripe = window.Stripe(window.stripePublishableKey);
+
+const getMessage = (code, default_message) => {
+  const map = {
+    resizing_machines_success: "Resizing successful! Reloading page...",
```
Comment (ClementChaumel) [inline static/js/src/advantage/manage-subscription.js:11]:
```suggestion
    resizing_machines_success: "blablabla something else",
```

Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:11]:
```suggestion
    resizing_machines_success: "Your changes were saved successfully, the page will reload",
```

File: static/js/src/advantage/manage-subscription.js [hunk 7]
```diff
@@ -1,93 +1,252 @@
...
+    cancelling_subscription_success:
+      "Subscription cancelled! Reloading page...",
```
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:13]:
```suggestion
      "Your subscription was cancelled, the page will reload",
```

File: static/js/src/advantage/manage-subscription.js [hunk 8]
```diff
@@ -1,93 +1,252 @@
...
+    resizing_machines_fail:
+      "<strong>Payment method:</strong> There was an error with the payment please <a href='/advantage/payment-methods'>update your payment methods</a> to fix it.",
```
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:15]:
```suggestion
      "<strong>Payment method:</strong> There was a problem with the payment please <a href='/advantage/payment-methods'>update your payment methods</a> to retry.",
```

File: static/js/src/advantage/manage-subscription.js [hunk 9]
```diff
@@ -1,93 +1,252 @@
...
+    subscription_missing:
+      "<strong>Cancelling subscription failed:</strong> It could be that you have a pending payment that is blocking this action. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
```
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:17]:
```suggestion
      "<strong>The subscription could not be cancelled:</strong> It could be that you have a pending payment that is blocking this action. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
```

File: static/js/src/advantage/manage-subscription.js [hunk 10]
```diff
@@ -1,93 +1,252 @@
...
+    cancelling_subscription_failed:
+      "<strong>Cancelling subscription failed:</strong> Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
```
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:19]:
```suggestion
      "<strong>The subscription could not be cancelled:</strong> Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
```

File: static/js/src/advantage/manage-subscription.js [hunk 11]
```diff
@@ -1,93 +1,252 @@
...
+    pending_purchase:
+      "<strong>Error:</strong> You already have a pending purchase. Please go to <a href='/advantage/payment-methods'>payment methods</a> to fix it. Reloading page...",
```
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:21]:
```suggestion
      "<strong>Error:</strong> You already have a pending purchase. Please go to <a href='/advantage/payment-methods'>payment methods</a> to retry.",
```

File: static/js/src/advantage/manage-subscription.js (hunk 1 above)
Comment (wgx) [reply static/js/src/advantage/manage-subscription.js:24]: done

File: static/js/src/advantage/manage-subscription.js (hunk 6 above)
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:11]:
```suggestion
    resizing_machines_success: "Your changes were saved successfully, the page will reload.",
```

File: static/js/src/advantage/manage-subscription.js (hunk 7 above)
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:13]:
```suggestion
      "Your subscription was cancelled, the page will reload.",
```

File: static/js/src/advantage/manage-subscription.js (hunk 8 above)
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:15]:
```suggestion
      "<strong>Payment method:</strong> There was problem with your payment please <a href='/advantage/payment-methods'>update your payment methods</a> to retry.",
```

File: static/js/src/advantage/manage-subscription.js (hunk 9 above)
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:17]:
```suggestion
      "<strong>Could not cancel subscription:</strong> It could be that you have a pending payment that is blocking this action. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.",
```

File: static/js/src/advantage/manage-subscription.js (hunk 11 above)
Comment (wgx) [inline static/js/src/advantage/manage-subscription.js:21]:
```suggestion
      "<strong>Error:</strong> You already have a pending purchase. Please go to <a href='/advantage/payment-methods'>payment methods</a> to retry.",
```

**Review by wgx — COMMENTED** [review]
pls make some copy changes @albertkol

**Comment by wgx** [comment]
UX+1

**Comment by ClementChaumel** [comment]
The issue with the wrong resizing is fixed but it still shows the wrong number of machines in the "machines" column of the table 
image

**Comment by albertkol** [comment]
@ClementChaumel yes. That is correct, sadly. We were requested that change as monthly products are not available anyway. 

The default value in the resize input field is still the old value that we used in the table row, as that is the actual number of monthly machines you have.

**Comment by wgx** [comment]
UX+1

**Review by ClementChaumel — APPROVED** [review]
:+1: you've got some conflicts though :)

**Review by ClementChaumel — APPROVED** [review]
LGTM!
